### ISSUE TITLE
Data resource for retrieving provider config

### ISSUE THREAD

**Issue by kerma** [terraform-providers/terraform-provider-postgresql#117]
### Terraform Version
Terraform v0.12.20
+ provider.postgresql v1.4.0

### Expected Behavior
There is a data source one could use to get provider configuration details.

### Actual Behavior
There is no data resource to get provider configuration details.

### Important Factoids
While developing a re-usable module I need to get access to the provider configuration, namely host and port values. Atm there is no other way to do that than to make the module user to pass in the values via variables. Dynamic provider definition is not very well supported by Terraform core, so I expect the provider definition to sit outside of this module. Which means I don't necessarily need to make the module user to pass those parameters into the module.

### References
https://github.com/hashicorp/terraform/issues/13018

**Comment by cyrilgdn** [comment]
Hi @kerma ,

Correct me if I'm wrong or if I misunderstood something but the provider configuration is not stored in the state so there's is no chance to have a data source for that. (And you can even change provider settings  but use the same Terraform state, there are not linked at all)

Not sure to get your use case, when creating the module, you don't need to pass the provider and you could (should) actually, as you mentioned, defined it outside of the module.

e.g.:
```hcl

$ cat module/db.tf
variable "db_name" {}

resource "postgresql_database" "db" {
  name = var.db_name
}

$ cat test.tf
provider "postgresql" {
  host     = "localhost"
  username = "postgres"
  password = "xxx"
}

module "database" {
  source = "./module"

  db_name = "test"
}

$ terraform init && terraform apply
Initializing modules...
- database in module

Initializing the backend...

Initializing provider plugins...
- Checking for available provider plugins...
- Downloading plugin for provider "postgresql" (terraform-providers/postgresql) 1.4.0...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, it is recommended to add version = "..." constraints to the
corresponding provider blocks in configuration, with the constraint strings
suggested below.

* provider.postgresql: version = "~> 1.4"

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.database.postgresql_database.db will be created
  + resource "postgresql_database" "db" {
      + allow_connections = true
      + connection_limit  = -1
      + encoding          = (known after apply)
      + id                = (known after apply)
      + is_template       = (known after apply)
      + lc_collate        = (known after apply)
      + lc_ctype          = (known after apply)
      + name              = "test"
      + owner             = (known after apply)
      + tablespace_name   = (known after apply)
      + template          = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.database.postgresql_database.db: Creating...
module.database.postgresql_database.db: Creation complete after 0s [id=test]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

And if needed, you can setup multiple instances of the same provider with [provider alias](https://www.terraform.io/docs/configuration/providers.html#alias-multiple-provider-instances).


I close this issue (as this is not doable), but feel free to open it back if I missed something or add share your use case/examples.

**Comment by kerma** [comment]
@cyrilgdn My use case is pretty much as I described. Think about a more complex module which takes care of many different things besides just Postgresql setup. For example:

1. I create a database, roles and all related configurations
2. I provision kubernetes resources
3. I set up AWS ALB

Let say I want to write down the database host name to the kubernetes configuration map. It would be nice to do this without making the module user to pass in this as a variable. 

As you mentioned I have all provider definitions outside of the module. It's already a lot of code to configure three different providers I force the module user to write. So reducing the amount of additional variables module user needs to set would make it better imo.

But yeah, I'm not sure it's really possible if the provider setup is not stored in the state. As it's probably very difficult to get the host definition from the active Postgres connection. I guess other providers rely on additional API's for that.
